docs/src/main/asciidoc/cdi.adoc

```diff
@@ -0,0 +1,361 @@
+
+== _Q: Nice. How does the dependency resolution work? I see no names or identifiers._
+
+A: That's a good question.
+In CDI the process of matching a bean to an injection point is *type-safe*.
+Each bean declares a set of bean types.
+In our example above, the `Translator` bean has two bean types: `Translator` and `java.lang.Object`.
+Subsequently, a bean is assignable to an injection point if the bean has a bean type that matches the _required type_ and has all the _required qualifiers_.
+We'll talk about qualifiers later.
+For now, it's enough to know that the bean above is assignable to an injection point of type `Translator` and `java.lang.Object`.
+
+== _Q: Hm, wait a minute. What happens if multiple beans declare the same type?_
+
+A: There is a simple rule: *exactly one bean must be assignable to an injection point, otherwise the build fails*.
+If none is assignable the build fails with `UnsatisfiedResolutionException`.
+If multiple are assignable the build fails with `AmbiguousResolutionException`.
+This is very useful because your application fails fast whenever the container is not able to find an unambiguous dependency for any injection point.
+
+[TIP]
+====
+Your can use programmatic lookup via  `javax.enterprise.inject.Instance` to resolve ambiguities at runtime and even iterate over all beans implementing a given type:
+
+[source,java]
+----
+public class Translator {
+
+    @Inject
+    Instance<Dictionary> dictionaries; <1>
+    
+    String translate(String sentence) { 
+      for (Dictionary dict : dictionaries) { <2>
+         // ...
+      }
+    }
+}
+----
+<1> This injection point will not result in an ambiguous dependency even if there are multiple beans that implement the `Dictionary` type.
+<2> `javax.enterprise.inject.Instance` extends `Iterable`.
+====
+
+== _Q: Can I use setter and constructor injection?_
+
+A: Yes, you can.
+In fact, in CDI the "setter injection" is superseded by more powerful https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#initializer_methods[initializer methods, window="_blank"].
+Intializers may accept multiple parameters and don't have to follow the JavaBean naming conventions.
+
+.Initialized and Constructor Injection Example
+[source,java]
+----
+@ApplicationScoped
+public class Translator {
+
+    private final TranslatorHelper helper
+
+    Translator(TranslatorHelper helper) { <1>
+       this.helper = helper;
+    }
+
+    @Inject <2>
+    void setDeps(Dictionary dic, LocalizationService locService) { <3>
+      / ...
+    }
+}
+----
+<1> This is a constructor injection. In fact, this code would not work in regular CDI implementations where a bean with a normal scope must always declare a no-args constructor. In Quarkus we detect this problem and "add" such a constructor directly in the bytecode.
+<2> An initializer method must be annotated with `@Inject`. 
+<3> An initializer may accept multiple parameters - each one is an injection point.
+
+== _Q: You talked about some qualifiers?_
```

Comment:
Should we also mention why would you want to use qualifiers? E.g. to distinguish between several beans having same type(s)?

Reply:
@manovotn isn't "Qualifiers are annotations that help the container to distinguish beans that implement the same type" clear enough?